File: src/me/seeking/event/Event.java
package me.seeking.event;

import me.seeking.Seeking;

import java.lang.reflect.InvocationTargetException;
import java.util.concurrent.CopyOnWriteArrayList;

public abstract class Event {

    private boolean cancelled;
    private Type type;

    public Event(Type type) {
        this.type = type;
        this.cancelled = false;
    }

    public enum Type {
        PRE, POST
    }

    public void call() {
        cancelled = false;

        CopyOnWriteArrayList<Data> dataList = Seeking.instance.eventManager.get(this.getClass());

        if (dataList == null)
            return;

        dataList.forEach(data -> {
            
            try {
                data.getTarget().invoke(data.getSource(), this);
            } catch (IllegalAccessException | InvocationTargetException e) {
                e.printStackTrace();
            }

        });


    }

    public Type getType() {
        return type;
    }

    public boolean isCancelled() {
        return cancelled;
    }

    public void setCancelled(boolean cancelled) {
        this.cancelled = cancelled;
    }


}

File: src/main/java/me/seeking/event/Event.kt
package me.seeking.event

import me.seeking.Seeking
import me.seeking.utils.PlayerUtil
import java.lang.reflect.InvocationTargetException
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer

abstract class Event(val type: Type) {
    var isCancelled = false

    enum class Type {
        PRE, POST
    }

    fun call() {
        isCancelled = false
        val dataList: CopyOnWriteArrayList<Data>? = Seeking.instance.eventManager!![this.javaClass]
        if (dataList != null) {
            if(dataList.isEmpty()){
                PlayerUtil.tellPlayer("NULL:"+this.javaClass.name)
                return
            }
        }
        dataList?.forEach(Consumer<Data> { data: Data ->
            try {
                data.target.invoke(data.source, this)
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            }
        })
    }
}